freqtrade/enums/liqformula.py

```diff
@@ -0,0 +1,86 @@
+from freqtrade.exceptions import OperationalException
+
+
+class LiqFormula(Enum):
+    """Equations to calculate liquidation price"""
+
+    BINANCE = "Binance"
+    KRAKEN = "Kraken"
+    FTX = "FTX"
+    NONE = None
+
+    def _exception(self, trading_mode: TradingMode, freq_specific: Optional[bool] = True):
+        """
+            Raises an exception if exchange used doesn't support desired leverage mode
+            :param trading_mode: cross, isolated, cross_futures or isolated_futures
+            :param freq_specific:
+                False if the exchange does not support this leverage mode
+                True if only freqtrade doesn't support it
+        """
+        if freq_specific:
+            raise OperationalException(
+                f"Freqtrade does not support {trading_mode.value} on {self.name}")
+        else:
+            raise OperationalException(f"{self.name} does not support {trading_mode.value} trading")
+
+    def _binance(self, trading_mode: TradingMode):
+        # TODO-lev: Additional arguments, fill in formulas
+
+        if trading_mode == TradingMode.CROSS_MARGIN:
+            # TODO-lev: perform a calculation based on this formula
+            # https://www.binance.com/en/support/faq/f6b010588e55413aa58b7d63ee0125ed
+            self._exception(trading_mode)
+        elif trading_mode == TradingMode.ISOLATED_MARGIN:
+            self._exception(trading_mode)  # Likely won't be implemented
+        elif trading_mode == TradingMode.CROSS_FUTURES:
+            # TODO-lev: perform a calculation based on this formula
+            # https://www.binance.com/en/support/faq/b3c689c1f50a44cabb3a84e663b81d93
+            self._exception(trading_mode)
+        elif trading_mode == TradingMode.ISOLATED_FUTURES:
+            # TODO-lev: perform a calculation based on this formula
+            # https://www.binance.com/en/support/faq/b3c689c1f50a44cabb3a84e663b81d93
+            self._exception(trading_mode)
+        else:
+            self._exception(trading_mode)
+
+    def _kraken(self, trading_mode: TradingMode):
+        # TODO-lev: Additional arguments, fill in formulas
+
+        if trading_mode == TradingMode.CROSS_MARGIN:
+            self._exception(trading_mode)
+            # TODO-lev: perform a calculation based on this formula
+            # https://support.kraken.com/hc/en-us/articles/203325763-Margin-Call-Level-and-Margin-Liquidation-Level
+        elif trading_mode == TradingMode.CROSS_FUTURES:
+            # TODO-lev: implement
+            self._exception(trading_mode)
+        elif trading_mode == TradingMode.ISOLATED_MARGIN or \
+                trading_mode == TradingMode.ISOLATED_FUTURES:
+            self._exception(trading_mode, True)
+        else:
+            self._exception(trading_mode)
+
+    def _ftx(self, trading_mode: TradingMode):
+        # TODO-lev: Additional arguments, fill in formulas
+        self._exception(trading_mode)
+
+    def __call__(self, **k):
```

Comment:
While this technically will for sure work - please keep enums as what they are - Lists of possibilities.

Implementation of the calculation should not go there (i've never evne seen it done this way in any other bigger project) - and the enum is for sure the last place i'd go look for a calculation.

Reply:
It's done this way so that sqlalchemy can store the value, like `InterestMode`

Reply:
Please don't mix 2 things here - SqlAlchemy is not related to what i mean above.
You can use an Enum just fine (we use that in other places too) - and use sqlalchemy will store the enum.


What i don't like is having Logic in the enum.
Sqlalchemy does not require you to convert the enum to a callable (implementing `__call__` converts it technically to a callable/function).
My point is to extract the LOGIC from the enum - so you don't call `InterestMode.Binance()` - but you pass it to a function as parameter(like `calc_interestmode(InterestMode.Binance)`.

The logic should go someplace else (not the `freqtrade.enums` package for sure) - maybe add a new Package for liquidation logic.
The enum (list of possible values) should stay in `freqtrade.enums` - while the logic should move.

This is also true for `InterestMode` (i allowed it temporarily there - but as i see now you're trying to use this too much, please do NOT store any logic in the enum). 

Also for `InterestMode` - we'll refactor this logic out fo the Enum before merging that to Develop.

Reply:
Where would `calc_interestmode` go?

Reply:
we'll need to be a bit careful with cyclic imports here - which is why i don't think we can (or should?) move it to a new file in `freqtrade.exchange` - as i've so far tried to avoid importing from this module to persistence - and it's another reason why i don't like it in enums - as importing things from "other" modules in `freqtrade.enums` can result in cyclic imports, too - depending where you need import from.

i'm not clear yet how many such calculations we're going to have ... if it's just 2 or 3 - we can have a file "somewhere" with it.
If there's "too many" - not sure what the limit should be -  maybe it makes sense to have them as a seperate module (`freqtrade.margin_calculations` or so) - which we can import where necessary.

Maybe start with having it as seperate file in `freqtrade.persistence`?
Moving a whole file from one module to another is usually a quick task (all we need to change is imports - as long as it's not something that's exposed to the user/strategy) - so i'm not too concerned about the precise location for now.